File: 백준/Silver/5397. 키로거/키로거.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.ListIterator;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int t = Integer.parseInt(br.readLine());

        for (int i = 0; i < t; i++) {
            String input = br.readLine();

            LinkedList<Character> list = new LinkedList<>();
            ListIterator<Character> it = list.listIterator();

            for (char ch : input.toCharArray()) {
                switch (ch) {
                    case '<':
                        if (it.hasPrevious()) {
                            it.previous();
                        }
                        break;
                    case '>':
                        if (it.hasNext()) {
                            it.next();
                        }
                        break;
                    case '-':
                        if (it.hasPrevious()) {
                            it.previous();
                            it.remove();
                        }
                        break;
                    default:
                        it.add(ch);
                        break;
                }
            }

            for (char ch : list) {
                sb.append(ch);
            }
            sb.append('\n');
        }
        System.out.print(sb);
    }
}

File: 백준/Silver/5397. 키로거/키로거.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.LinkedList

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val test = readLine().toInt()
    repeat(test) {
        val str = readLine().toString()
        val list = LinkedList<Char>()
        var cursor = 0
        for (i in str) {
            when (i) {
                '<' -> {
                    if (cursor != 0) {
                        cursor--
                    }
                }
                '>' -> {
                    if (cursor < list.size) {
                        cursor++
                    }
                }
                '-' -> {
                    if (cursor != 0 && list.isNotEmpty()) {
                        list.removeAt(--cursor)
                    }
                }
                else -> {
                    list.add(cursor++, i)
                }
            }
        }
        println(list.toCharArray())
    }
}